### DISCUSSION TITLE
Is it possible to click trigger submenu, not hover?

### DISCUSSION THREAD

**Discussion by guijiangheng** [radix-ui/primitives#1187] [Ideas]
Currently submenu is opened by hover TriggerItem, but on some occasion, for example click show more button to expand submenu is preferred. Is this behavior supported?

**Comment by guijiangheng** [comment]
<img width="231" alt="截屏2022-02-24 下午3 16 02" src="https://user-images.githubusercontent.com/24951305/155476419-c795e9b6-c255-45dc-9555-e1161f42b19f.png">
Just like this UI design

**Reply by jjenzz** [reply]
Are you wanting a menu to open when user clicks the `...` or `Default View`?

@andy-hook @benoitgrelard If the former, I'm wondering if this could be another case where we could consider exposing `Menu.Sub` as a separate part so that people can create menus within menus, instead of submenus.

I'm not sure of the a11y implications though...

**Reply by benoitgrelard** [reply]
Yeah I'm not sure how that would be accessible still… How would you navigate it keyboard-wise? It wouldn't work as a single tab-stop anymore…

**Comment by icoloma** [comment]
I have a similar issue. I would like this submenu (vertical layout, submenus appear under the selected menu item) to display when I click on "Tickets" instead of when I hover. Hovering in this case is super distracting, because the options jump up and down when you move the cursor through the menu items (keyboard works fine).

I know that this is not the typical layout (`Sub` is displayed under the `Trigger`, not to the right), but is `NavigationMenu` the wrong component for this? 

**EDIT** It seems that this case is not that different from the index in the documentation for radix-ui, so maybe I'm overdesigning and should just sprinkle some aria roles over links and `navs`

![image](https://user-images.githubusercontent.com/222529/163815248-9f1c6dd0-e09c-4b11-9860-e3eb601955d1.png)

**Reply by benoitgrelard** [reply]
Hey @guijiangheng I'm not sure what you are building here, but this doesn't look like a `DropdownMenu` which is the component discussed originally here.

**Reply by icoloma** [reply]
Damn. Sorry, my bad. I didn't notice that this was indeed a different component. Apologies. 

Anyway, I solved it outside of radix-ui, the navigation was not hard to implement. Thanks for an awesome library!

**Reply by fera2k** [reply]
@icoloma How did you solve it? Would you care to share? I'm facing a similar issue.

We're creating a responsible navBar using the `NavigationMenu`, that collapses to a sidebar when the resolution is small (phone like). With mobile emulator is ok, because it doesn't trigger the "on mouse over". But when still using the actual mouse, in a window with smaller width, I really would like to disable the `mouseover` and only show the sub items if the user clicks on the trigger. Any idea how to get this to work?

**Reply by vitharanagedonjani-i2e** [reply]
@icoloma How did you solve it?

**Reply by icoloma** [reply]
What I really wanted was a set of labels for some menu groups, and keep them always open:

```html
<nav aria-label="Primary" class="sidebar-menu-nav">
  <ul class="menu-contents">
    <li class="menu-item-li">
      <a class="menu-item" href="/link1">Edit</a>
    </li>
    <label id="341675719" class="menu-group-label">Tickets</label>
    <ul role="group" class="menu-group" aria-labelledby="341675719">
      <li class="menu-item-li">
        <a class="menu-item" href="/link21">Tickets sold</a>
      </li>
      <li class="menu-item-li">
        <a class="menu-item" href="/link22">Classes</a>
      </li>
    </ul>
  </ul>
</nav>
```

![image](https://user-images.githubusercontent.com/222529/194638473-21d3b83b-8d52-42b2-b996-369ba9470081.png)


We were inspired by the [Kubernetic UI](https://www.kubernetic.com/blog/release-3-1-0) IIRC

**Reply by vitharanagedonjani-i2e** [reply]
@icoloma I still need to disabled this on hover and let it work only on click, do you think there is a workaround or it won't work?

**Reply by benoitgrelard** [reply]
@vitharanagedonjani-i2e you could try overriding our event handling and implementing your own whilst controlling the `open` state for submenus.

**Reply by vitharanagedonjani-i2e** [reply]
@benoitgrelard could you demonstrate by any chance?

**Reply by benoitgrelard** [reply]
Would have to look at implementation details, but basic steps would be:
- On `SubTrigger`, `event.preventDefault()` in the handlers we use for pointer stuff (from memory `onPointerMove` and `onPointerLeave`)
- Control `open` state of `Sub` using `open` and `onOpenChange` props
- set the `open` state accordingly to your needs (your new event handling on `SubTrigger`)

I can't guarantee everything will work smoothly as the component hasn't been designed to work this way and there's a bunch of stuff happening to handle hover interactions specifically but generally speaking, you should be able to override our work, although this isn't something we would recommend.

Hope that helps.

**Reply by vitharanagedonjani-i2e** [reply]
@benoitgrelard I tried this and it worked just fine. Only issue is in submenus when I `event.preventDefault()` the pointer events of the trigger it opens on click but it doesn't close afterwards unless I close the main trigger which opened the sub level. Is this the default behaviour of a submenu? Also I couldn't find a prop called `onOpenChange` in the codebase.

[](https://www.awesomescreenshot.com/video/11649982?key=62a971488b2f725a2355d5f64fb067f6)

**Reply by benoitgrelard** [reply]
> Only issue is in submenus when I event.preventDefault() the pointer events of the trigger it opens on click but it doesn't close afterwards unless I close the main trigger which opened the sub level

Not sure I follow sorry.

> Also I couldn't find a prop called onOpenChange in the codebase.

<img width="829" alt="CleanShot 2022-10-11 at 14 47 15@2x" src="https://user-images.githubusercontent.com/1539897/195108654-17381bd0-97b6-42de-b83b-dd1a928619e0.png">

**Reply by vitharanagedonjani-i2e** [reply]
@benoitgrelard  https://www.radix-ui.com/docs/primitives/components/navigation-menu#sub Could you take a look at this link?

**Reply by benoitgrelard** [reply]
I assumed you meant `DropdownMenu`, are you trying to do this with `NavigationMenu` this whole time?

**Reply by vitharanagedonjani-i2e** [reply]
@benoitgrelard yes, I think there must have been a miss understanding. However I manage to fix this using the value property. But still I need the submenu to be locked in "open" state on initial trigger. Is there any suggestion on how to fix that?

**Reply by benoitgrelard** [reply]
Sorry still not really following, maybe build a codesandbox example?

**Reply by vitharanagedonjani-i2e** [reply]
https://codesandbox.io/s/radix-navigation-to-keep-the-submenu-open-36ogr4?file=/pages/navigation.tsx

@benoitgrelard  So here I got 2 menu items and I want keep them at `open` state after I click on any one of them. (Shouldn't be able close)

**Reply by benoitgrelard** [reply]
It looks like there's a bunch of code still running on pointer enter/leave potentially.
Can I ask you what you're trying to build though? This seems like a weird interaction.

**Reply by vitharanagedonjani-i2e** [reply]
@benoitgrelard  I'm trying to build a SideNav, Which opens the menu items on click(not hover; that's why i'm using the pointer events to disable the hover effect). Also there are places where menu items has sub menus and I use `NavigationMenu.Sub` for that. I need to preserve them at `[data-state="open"]` after being opened. Because I'm using an external button to close them. Currently even if I disabled the `pointerEvents` they are getting closed on click

**Reply by djalmaaraujo** [reply]
@benoitgrelard I am facing the same situation. Using Navigation Menu to build a sidebar, where usually you click to expand items, not use the hover. Any ideas?

**Reply by katannshaw** [reply]
I had the same issue and found two approaches in other locations and they both work for me. I decided to use the regular function method.

Using inline function calls (see line 325)
https://codesandbox.io/s/unruffled-robinson-duwvgn?file=/App.js

Using regular function
https://github.com/radix-ui/primitives/issues/1630#issuecomment-1491288625

Hope it helps.

**Comment by ofelipevaz** [comment]
I solved this problem by using the **stopPropagation** method on the click event.

```tsx
<DropdownMenu.Sub open={open}> {/* use "open" in useState */}
    <DropdownMenu.SubTrigger
      onClick={(e) => {
        e.stopPropagation() // prevent menu item select when click
        setOpen(!open)
      }}
     >
       ...
    </DropdownMenu.SubTrigger>
</DropdownMenu.Sub>
```
